### PR TITLE
Entrypoint hooks (SYN-4250)

### PR DESCRIPTION
**Pull request by vEpiphyte**

### CODE REVIEW

File: docs/synapse/devguides/certbot.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
+# Certbot preboot example
+# Author: william.gibb@vertex.link
+
+# This script is an example of using Let's Encrypt certbot tool to generate
+# a valid HTTPS certificate for a Synapse service. This example uses web
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:7]: "This example uses web"?

File: docs/synapse/devguides/certbot.sh [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#
+# This creates and stores a Python venv in the
+# /vertex/storage/preboot/letsencrypt/venv directory, so the certbot
+# tool it installed once in a separate python environment, and cached in
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:11]: typo "it" instead of "is"

File: docs/synapse/devguides/certbot.sh [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+# a mapped volume.
+#
+# Once the venv is setup, certbot is is used to create and potentially renew
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:14]: "is is"

File: docs/synapse/devguides/certbot.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+# the locations in /vertex/storage where Synapse services assume they will
+# find the HTTPS keys.
+#
+# cerbot does use a random backoff timer when performing a renewal. There may
+# be a random delay when starting a service when the certificate needs to be
+# renewed.
+#
+# Required Environment variables:
+#
+# CERTBOT_HOSTNAME - the hostname that certbot will generate certificate for.
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:25]: generate "a" certificate for

File: docs/synapse/devguides/certbot.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+# /vertex/storage/preboot/letsencrypt/venv directory, so the certbot
+# tool it installed once in a separate python environment, and cached in
+# a mapped volume.
+#
+# Once the venv is setup, certbot is is used to create and potentially renew
+# an HTTPS certificate. This certificate and private key are then copied to
+# the locations in /vertex/storage where Synapse services assume they will
+# find the HTTPS keys.
+#
+# cerbot does use a random backoff timer when performing a renewal. There may
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:19]: "cerbot"

File: docs/synapse/devguides/certbot.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+CONFDIR=$LEDIR/conf
+LOGSDIR=$LEDIR/logs
+WORKDIR=$LEDIR/work
+VENV=$LEDIR/venv
+
+mkdir -p $LOGSDIR
+mkdir -p $CONFDIR
+mkdir -p $WORKDIR
+
+CERBOT_DIR_ARGS=" --work-dir ${WORKDIR} --logs-dir=${LOGSDIR} --config-dir=${CONFDIR} "
```
Comment (Cisphyx) [inline docs/synapse/devguides/certbot.sh:62]: missing T in "CERBOT_DIR_ARGS", used a couple times

File: docs/synapse/devopsguide.rst [hunk 7]
```diff
@@ -589,6 +588,60 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
 
 This would indicate the use of a deprecated API.
 
+Entrypoint Hooking
+------------------
+
+Synapse service containers provide two ways that users can modify the container startup process, in order to execute
+their own scripts or commands.
+
+The first way is using a script that executes before services start can be configured set by mapping in a file at
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:597]: missing a word or two here i think

File: docs/synapse/devopsguide.rst [hunk 8]
```diff
@@ -589,6 +588,60 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
...
+``/vertex/preboot/run`` and making sure it is marked as an executable. If this file is present, the file will be
+executed prior to booting the service. If this executable does not have return ``0``, the container will fail to start
+up.
+
+One example for using this hook is to use ``certbot`` to create HTTPS certificates for a Synapse service. This example
+assume the Cortex is running as root with, so that certbot can bind port 80 to perform the ``http-01`` challenge.
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:603]: assumes

File: docs/synapse/devopsguide.rst [hunk 9]
```diff
@@ -589,6 +588,60 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
...
+Create a preboot directory::
+
+  mkdir -p /srv/syn/le.cortex/preboot
+
+Copy the following script to ``/vertex/preboot/run`` and use ``chmod`` to mark it as an executable file:
+
+.. literalinclude:: devguides/certbot.sh
+    :language: bash
+
+That volume be mounted at ``/vertex/preboot``. The following docker-compose file shows mounting that directory into
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:614]: lol, kinda works but "That volume be mounted at"

File: docs/synapse/devopsguide.rst [hunk 10]
```diff
@@ -589,6 +588,60 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
...
+      environment:
+          SYN_LOG_LEVEL: "DEBUG"
+          SYN_CORTEX_STORM_LOG: "true"
+          CERTBOT_HOSTNAME: "cortex.acme.corp"
+          CERTBOT_EMAIL: "user@acme.corp"
+
+When started, the container will attempt to run the script before starting the Cortex service.
+
+The second way is running a concurrent script can be set by mapping in a file at ``/vertex/concurrent/run``, also as
+an executable file. If this file is present, the file will be executed as a backgrounded task. This script executed
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:635]: "is" executed

File: docs/synapse/devopsguide.rst [hunk 11]
```diff
@@ -589,6 +588,60 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
...
+prior to starting up the Synapse service. This executable would be stopped when the container is stopped.
+
+.. note::
+
+    If a volume is mapped into ``/vertex/preboot/`` or ``/vertex/concurrent/``, it will not be included in any backups
+    made by a Synapse service using the backup APIs. Making backups of any data persisted in these locations is
+    the responsibility of operator configuring the container.
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:642]: "the" operator

File: docs/synapse/devopsguide.rst [hunk 12]
```diff
@@ -589,6 +588,61 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
...
+Synapse service containers provide two ways that users can modify the container startup process, in order to execute
+their own scripts or commands.
+
+The first way modify the startup process is using a script that executes before services start. This can be configured
+by mapping in a file at ``/vertex/preboot/run`` and making sure it is marked as an executable. If this file is present,
+the file will be executed prior to booting the service. If this executable does not have return ``0``, the container
+will fail to start up.
+
+One example for using this hook is to use ``certbot`` to create HTTPS certificates for a Synapse service. This example
+assumes the Cortex is running as root with, so that certbot can bind port 80 to perform the ``http-01`` challenge.
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:603]: "running as root with"

File: docs/synapse/devopsguide.rst [hunk 13]
```diff
@@ -589,6 +588,61 @@ if a remote caller uses the ``eval()`` API on a Cortex, it would log the followi
 
 This would indicate the use of a deprecated API.
 
+Entrypoint Hooking
+------------------
+
+Synapse service containers provide two ways that users can modify the container startup process, in order to execute
+their own scripts or commands.
+
+The first way modify the startup process is using a script that executes before services start. This can be configured
```
Comment (Cisphyx) [inline docs/synapse/devopsguide.rst:597]: first way "to" modify

**Review by invisig0th — APPROVED** [review]

**Pull request merged**